contracts/reserve/ExchangeHelper.sol

```diff
@@ -0,0 +1,340 @@
+			contributionAmount,
+			result,
+			receiver
+		);
+		return result;
+	}
+
+	/**
+	 * @dev Redeem cDAI to DAI
+	 * @param _amount Amount of cDAI to redeem for DAI
+	 * @return the amount of DAI received
+	 */
+	function _redeemDAI(uint256 _amount) internal returns (uint256) {
+		cERC20 cDai = cERC20(cDaiAddress);
+		ERC20 dai = ERC20(daiAddress);
+
+		uint256 currDaiBalance = dai.balanceOf(address(this));
+
+		uint256 daiResult = cDai.redeem(_amount);
+		require(daiResult == 0, "cDai redeem failed");
+
+		uint256 daiReturnAmount = dai.balanceOf(address(this)) - currDaiBalance;
+
+		return daiReturnAmount;
+	}
+
+	/**
+	 * @dev Convert Dai to CDAI and buy
+	 * @param _amount DAI amount to convert
+	 * @param _minReturn The minimum allowed return in GD tokens
+	 * @param _targetAddress address of g$ and gdx recipient if different than msg.sender
+	 * @return (gdReturn) How much GD tokens were transferred
+	 */
+	function _cdaiMintAndBuy(
+		uint256 _amount,
+		uint256 _minReturn,
+		address _targetAddress
+	) internal returns (uint256) {
+		GoodReserveCDai reserve = GoodReserveCDai(
+			nameService.getAddress("RESERVE")
+		);
+		cERC20 cDai = cERC20(cDaiAddress);
+
+		uint256 currCDaiBalance = cDai.balanceOf(address(this));
+
+		//Mint cDAIs
+		uint256 cDaiResult = cDai.mint(_amount);
+		require(cDaiResult == 0, "Minting cDai failed");
+
+		uint256 cDaiInput = cDai.balanceOf(address(this)) - currCDaiBalance;
+		cDai.transfer(address(reserve), cDaiInput);
+		return reserve.buy(cDaiInput, _minReturn, _targetAddress);
+	}
+
+	/**
+	@dev Helper to swap tokens in the Uniswap
+	*@param _token token to buy or sell for
+	*@param _tokenAmount token amount to sell or buy
+	*@param _isBuy if swap transaction to buy DAI or sell DAI
+	*@param _minDAIAmount minimum DAI amount to get in swap transaction if transaction is buy
+	*@param _minTokenReturn minimum token amount to get in swap transaction if transaction is sell
+	*@param _receiver receiver of tokens after swap transaction
+	 */
+	function _uniswapSwap(
+		ERC20 _token,
+		uint256 _tokenAmount,
+		bool _isBuy,
+		uint256 _minDAIAmount,
+		uint256 _minTokenReturn,
+		address _receiver
+	) internal returns (uint256[] memory) {
+		Uniswap uniswapContract = Uniswap(
+			nameService.getAddress("UNISWAP_ROUTER")
+		);
+		address[] memory path = new address[](2);
+
+		uint256[] memory swap;
+		if (address(_token) == address(0x0)) {
+			if (_isBuy) {
+				path[0] = uniswapContract.WETH();
+				path[1] = daiAddress;
+				swap = uniswapContract.swapExactETHForTokens{
+					value: _tokenAmount
+				}(_minDAIAmount, path, address(this), block.timestamp);
+				return swap;
+			} else {
+				path[0] = daiAddress;
+				path[1] = uniswapContract.WETH();
+				swap = uniswapContract.swapExactTokensForETH(
+					_tokenAmount,
+					_minTokenReturn,
+					path,
+					_receiver,
+					block.timestamp
+				);
+				return swap;
+			}
+		} else {
+			if (_isBuy) _token.approve(address(uniswapContract), _tokenAmount);
```

Comment:
sell doesnt reuqire approve? what's the difference
dont use _isBuy, simply use inputToken and outputToken as inputs to the helper
path[0]=inputtoken
path[1]=outputtoken
if(path[0]==eth)
swapexactEth
else if(path[1]) == eth
swap exacttokenforeth
else
{
approve
swapExactTokensForTokens
}

Reply:
"sell doesnt reuqire approve? what's the difference"
Sell is always dai to the tokens which is desired so when we approve infinite amount of dai to spend uniswap router then it's fine

Reply:
changed @sirpy